src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+            // https://docs.microsoft.com/en-us/visualstudio/ide/code-snippets-schema-reference?view=vs-2019
+            return new XDocument(
+                new XDeclaration("1.0", "utf-8", null),
+                new XElement(
+                    snippetNamespace + "CodeSnippets",
+                    new XElement(
+                        snippetNamespace + "CodeSnippet",
+                        new XAttribute(snippetNamespace + "Format", "1.0.0"),
+                        new XElement(
+                            snippetNamespace + "Header",
+                            new XElement(
+                                snippetNamespace + "SnippetTypes",
+                                new XElement(snippetNamespace + "SnippetType", new XText("Expansion"))),
+                            new XElement(snippetNamespace + "Title", new XText(methodName)),
+                            new XElement(snippetNamespace + "Author", "Microsoft"),
+                            new XElement(snippetNamespace + "Description"),
+                            new XElement(snippetNamespace + "HelpUrl"),
+                            new XElement(snippetNamespace + "Shortcut", methodName)),
+                        new XElement(
+                            snippetNamespace + "Snippet",
+                            new XElement(snippetNamespace + "Declarations", declarations.ToArray()),
+                            new XElement(
+                                snippetNamespace + "Code",
+                                new XAttribute(snippetNamespace + "Language", "csharp"),
+                                new XCData(template.ToString()))))));
+        }
+
+        private void OnModelUpdated(object sender, Model e)
+        {
+            AssertIsForeground();
+
+            if (e is null)
+            {
+                return;
+            }
+
+            if (!_state.IsFullMethodCallSnippet)
+            {
+                return;
+            }
+
+            var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document is null)
+            {
+                return;
+            }
+
+            var newSymbolKey = (e.SelectedItem as AbstractSignatureHelpProvider.SymbolKeySignatureHelpItem)?.SymbolKey ?? default;
+            var newSymbol = newSymbolKey.Resolve(document.Project.GetRequiredCompilationAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None), cancellationToken: CancellationToken.None).GetAnySymbol();
+            if (newSymbol is not IMethodSymbol method)
+                return;
+
+            MoveToSymbol(_state._symbols, method, CancellationToken.None);
+        }
+
+        private async Task<ImmutableArray<ISymbol>> GetSymbolsAsync(
+            SnapshotPoint caretPosition,
+            CancellationToken cancellationToken)
+        {
+            var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
don't love that you're doing this in an async method.  could you move this *before* the JTF call?

Reply:
⏱ promoted to parameter